### PR TITLE
CDAP-15844: Kudu plugin validation.

### PR DESCRIPTION
**Pull request by AntonGensitskiy**
JIRA: https://issues.cask.co/browse/CDAP-15844
WIKI: https://wiki.cask.co/display/CE/Guidelines+on+Plugin+Validation

In scope of this PR:

- integrated with new validation API

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -29,15 +29,12 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <spark.version>2.1.3</spark.version>
-    <cdap.version>6.0.0-SNAPSHOT</cdap.version>
+    <cdap.version>6.1.0-SNAPSHOT</cdap.version>
     <hydrator.version>2.2.0-SNAPSHOT</hydrator.version>
     <hadoop.version>2.3.0</hadoop.version>
     <kudu.version>1.2.0-cdh5.10.0</kudu.version>
     <widgets.dir>widgets</widgets.dir>
```
Comment (aonischuk) [inline pom.xml]: remove

File: pom.xml [hunk 2]
```diff
@@ -29,15 +29,12 @@
...
     <docs.dir>docs</docs.dir>
```
Comment (aonischuk) [inline pom.xml]: same

File: pom.xml [hunk 3]
```diff
@@ -29,15 +29,12 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <spark.version>2.1.3</spark.version>
-    <cdap.version>6.0.0-SNAPSHOT</cdap.version>
+    <cdap.version>6.1.0-SNAPSHOT</cdap.version>
     <hydrator.version>2.2.0-SNAPSHOT</hydrator.version>
```
Comment (aonischuk) [inline pom.xml]: 2.3.0-SNAPSHOT

File: pom.xml [hunk 4]
```diff
@@ -29,15 +29,12 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Comment (aonischuk) [inline pom.xml:30]: Please add checkstyle and rat plugins.
Also suppressions.xml and checkstyle.xml files.

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java [hunk 5]
```diff
@@ -94,18 +93,13 @@ public KuduSink(KuduSinkConfig kuduSinkConfig) {
   @Override
   public void configurePipeline(PipelineConfigurer configurer) {
     super.configurePipeline(configurer);
-    Preconditions.checkArgument(!Strings.isNullOrEmpty(kuduSinkConfig.optSchema), "Write schema is not specified. Please add" +
-      "the write schema.");
+    FailureCollector failureCollector = configurer.getStageConfigurer().getFailureCollector();
+    kuduSinkConfig.validate(failureCollector);
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSink.java:97]: do we need `failureCollector.getOrThrowException()` here?

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java [hunk 6]
```diff
@@ -94,18 +93,13 @@ public KuduSink(KuduSinkConfig kuduSinkConfig) {
...
 
     // Checks if that we are writing with has been constructed correctly.
     Schema writeSchema = kuduSinkConfig.getSchema();
     configurer.getStageConfigurer().setOutputSchema(writeSchema);
 
-    // If there is macro specified for 'master' address or table name, then
-    // we defer the creation of table to initialize.
-    if (kuduSinkConfig.containsMacro("master") || kuduSinkConfig.containsMacro("name")) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSink.java:106]: Why did we remove this again? Won't `createKuduTable` fail when those configs contain macro?

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java [hunk 7]
```diff
@@ -260,4 +317,139 @@ public int getSeed() {
   public int getThreads() {
     return (optBossThreads != null) ? Integer.parseInt(optBossThreads) : 1;
   }
+
+  public void validate(FailureCollector failureCollector) {
+    if (Strings.isNullOrEmpty(optSchema)) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: Control this with @Nullable. No checks for empty values should be present. Nullable annotation does it for us. As far as I understand it.

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java [hunk 8]
```diff
@@ -260,4 +317,139 @@ public int getSeed() {
...
+                                     "Provide correct schema.")
+          .withStacktrace(e.getStackTrace())
+          .withConfigProperty(SCHEMA);
+      }
+    }
+
+    // If there is macro specified for 'master' address or table name, then
+    // we defer the creation of table to initialize.
+    if (!containsMacro(MASTER)) {
+      if (Strings.isNullOrEmpty(optMasterAddresses)) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: remove this check. We should control the value being present with adding or not adding @nullable.

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java [hunk 9]
```diff
@@ -260,4 +317,139 @@ public int getSeed() {
...
+        failureCollector.addFailure("Kudu Master Server address list is empty.",
+                                    "Provide Kudu Master Server address list.")
+          .withConfigProperty(MASTER);
+      } else {
+        Arrays.stream(optMasterAddresses.split(",")).forEach(masterAddress -> {
+          if (masterAddress.split(":").length != 2){
+            failureCollector.addFailure(String.format("Master Address '%s' has incorrect format.", masterAddress),
+                                        "Provide Master Address in format - <hostname>:<port>")
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: shouldn't it be `<hostname1>:<port1>,<hostname2>:<port2>,...`?

Also why some words are capitalized?

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java [hunk 10]
```diff
@@ -260,4 +317,139 @@ public int getSeed() {
...
+              .withConfigProperty(MASTER);
+          }
+        });
+      }
+    }
+
+    if (!containsMacro(NAME) && Strings.isNullOrEmpty(optTableName)) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: remove this checks. We should control with with adding or not adding @nullable.

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java [hunk 11]
```diff
@@ -16,67 +16,152 @@
...
+  public static final String SCHEMA = "schema";
+  public static final String COLUMNS = "columns";
+  public static final String OPT_TIMEOUT = "opt-timeout";
 
-  @Name("referenceName")
+  // Required Fields.
+  @Name(REFERENCE_NAME)
   @Description("This will be used to uniquely identify this source/sink for lineage, annotating metadata, etc.")
-  public String referenceName;
+  private String referenceName;
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java:45]: Please validate referenceName.
`IdUtils.validateReferenceName(referenceName, failureCollector)`

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java [hunk 12]
```diff
@@ -16,67 +16,152 @@
...
    *
    * @throws IllegalArgumentException throw when there is issue with the configuration.
    */
-  public void validate() throws IllegalArgumentException {
-    if (!containsMacro("master")) {
-      Preconditions.checkArgument(!Strings.isNullOrEmpty(optMasterAddresses),
-                                  "Kudu Master Server address list is empty.");
+  public void validate(FailureCollector failureCollector) throws IllegalArgumentException {
+    if (!containsMacro(MASTER)){
+      if (Strings.isNullOrEmpty(optMasterAddresses)) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: remove these checks. Add or not add @nullable instead to control this.

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java [hunk 13]
```diff
@@ -16,67 +16,152 @@
...
+        failureCollector.addFailure("Kudu Master Server address list is empty.",
+                                    "Provide Kudu Master Server address list.")
+          .withConfigProperty(MASTER);
+      } else {
+        Arrays.stream(optMasterAddresses.split(",")).forEach(masterAddress -> {
+          if (masterAddress.split(":").length != 2){
+            failureCollector.addFailure(String.format("Master Address '%s' has incorrect format.", masterAddress),
+                                        "Provide Master Address in format - <hostname>:<port>")
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: `<hostname1>:<port1>,<hostname2>:<port2>,...` ?

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java [hunk 14]
```diff
@@ -16,67 +16,152 @@
...
+    try {
+      Schema.parseJson(optSchema);
+    } catch (IOException e) {
+      failureCollector.addFailure( String.format("Unable to parse schema '%s'", optSchema),
+                                   "Provide correct schema.")
+        .withStacktrace(e.getStackTrace())
+        .withConfigProperty(SCHEMA);
+    }
+
+    if (!containsMacro(NAME) && Strings.isNullOrEmpty(optTableName)) {
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: same

**Review by aonischuk — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java [hunk 15]
```diff
@@ -94,18 +93,13 @@ public KuduSink(KuduSinkConfig kuduSinkConfig) {
...
-      return;
-    }
     createKuduTable();
```
Comment (aonischuk) [inline src/main/java/io/cdap/plugin/kudu/KuduSink.java]: Can we validate that this method doesn't fail as well?

File: pom.xml (hunk 1 above)
Comment (AntonGensitskiy) [reply pom.xml]: Done

File: pom.xml (hunk 2 above)
Comment (AntonGensitskiy) [reply pom.xml]: Done.

File: pom.xml (hunk 3 above)
Comment (AntonGensitskiy) [reply pom.xml]: Done.

File: pom.xml (hunk 4 above)
Comment (AntonGensitskiy) [reply pom.xml:30]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java (hunk 5 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSink.java:97]: add failureCollector.getOrThrowException()

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java (hunk 6 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSink.java:106]: I rollback this change

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java (hunk 7 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java (hunk 8 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java (hunk 9 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: `<hostname1>:<port1>,<hostname2>:<port2>,...` correct format
we check that each master address should be in format `<hostname1>:<port1>`

File: src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java (hunk 10 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSinkConfig.java]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java (hunk 11 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java:45]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java (hunk 12 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java (hunk 13 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: `<hostname1>:<port1>,<hostname2>:<port2>,... `correct format
we check that each master address should be in format `<hostname1>:<port1>`

File: src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java (hunk 14 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/kudu/KuduSink.java (hunk 15 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/kudu/KuduSink.java]: Added validation check

**Pull request merged**